main.go

```diff
@@ -3,117 +3,185 @@ package main
+	if err != nil {
+		logger.Fatal(err)
+	}
 
-	sigs := make(chan os.Signal, 1)
-	signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
+	threadFinishedCh := make(chan struct{}, 2)
+	srv, err := initialiseExtension(logger, threadFinishedCh)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	shutdownChannel := make(chan struct{})
 	go func() {
-		s := <-sigs
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, syscall.SIGTERM, syscall.SIGINT)
+		select {
+		case s := <-interruptChannel:
+			logger.Printf("Received %s, exiting\n", s)
+		case <-shutdownChannel:
+			logger.Println("Received shutdown event, exiting")
+		}
+
 		cancel()
-		logger.Println("Received", s)
-		logger.Println("Exiting")
+		if err := srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners, or context timeout:
+			logger.Printf("HTTP server shutdown error: %s\n", err)
+		}
+		threadFinishedCh <- struct{}{}
 	}()
 
-	_, err := extensionClient.Register(ctx, extensionName)
-	if err != nil {
-		logger.Fatal(err)
-	}
+	processEvents(ctx, logger, extensionClient)
 
-	initialiseExtension(logger)
+	// Once processEvents returns, signal that it's time to shutdown.
+	shutdownChannel <- struct{}{}
 
-	processEvents(ctx, logger)
+	// Ensure we wait for the HTTP server to gracefully shut down.
+	// One signal from the HTTP server thread and one from the shutdown thread.
+	<-threadFinishedCh
+	<-threadFinishedCh
+	logger.Println("Graceful shutdown complete")
 }
 
-func initialiseExtension(logger *log.Logger) {
+func initialiseExtension(logger *log.Logger, threadFinishedCh chan struct{}) (*http.Server, error) {
 	logger.Println("Initialising")
 
 	vaultAddr := os.Getenv("VAULT_ADDR")
 	vaultAuthRole := os.Getenv("VAULT_AUTH_ROLE")
 	vaultAuthProvider := os.Getenv("VAULT_AUTH_PROVIDER")
-	vaultIamServerId := os.Getenv("VAULT_IAM_SERVER_ID")	// Optional
+	vaultIAMServerID := os.Getenv("VAULT_IAM_SERVER_ID") // Optional
 
-	configuredSecrets, err := config.ParseConfiguredSecrets()
+	if vaultAddr == "" || vaultAuthProvider == "" || vaultAuthRole == "" {
+		return nil, errors.New("missing VAULT_ADDR, VAULT_AUTH_PROVIDER or VAULT_AUTH_ROLE environment variables")
+	}
+
+	client, err := vault.NewClient(logger, vaultAuthRole, vaultAuthProvider, vaultIAMServerID)
 	if err != nil {
-		logger.Fatalf("Failed to parse configured secrets to read: %s", err)
+		return nil, fmt.Errorf("error getting client: %s", err)
+	} else if client == nil {
+		return nil, fmt.Errorf("nil client returned: %s", err)
 	}
 
-	if vaultAddr == "" || vaultAuthProvider == "" || vaultAuthRole == "" || len(configuredSecrets) == 0 {
-		logger.Fatal("missing VAULT_ADDR, VAULT_AUTH_PROVIDER, VAULT_AUTH_ROLE, or VAULT_SECRET_ environment variables.")
+	err = writePreconfiguredSecrets(client)
+	if err != nil {
+		return nil, err
 	}
 
-	client, err := vault.NewClient(logger, vaultAuthRole, vaultAuthProvider, vaultIamServerId)
+	srv := server.New(client)
```

Comment:
I think we're missing some inputs here? 

```
./main.go:98:19: not enough arguments in call to server.New
        have (*api.Client)
        want (*log.Logger, *api.Client)
```

Reply:
Oops not sure how that happened. Should be fixed in the latest commits.